dep0/lib/05_llvmgen/src/gen.cpp

```diff
@@ -178,45 +163,43 @@ static void gen_func_body(
     llvm::Function*);
 
 static bool is_first_order_abstraction(typecheck::expr_t::abs_t const&);
-static bool is_first_order_application(typecheck::expr_t::app_t const&);
-static bool is_first_order_type(typecheck::type_t const&);
-static bool is_first_order_function_type(typecheck::type_t::arr_t const&);
+static bool is_first_order_type(typecheck::expr_t const&);
+static bool is_first_order_function_type(typecheck::expr_t::pi_t const&);
 
 bool is_first_order_abstraction(typecheck::expr_t::abs_t const& f)
 {
-    return is_first_order_type(f.ret_type) and
-        std::ranges::all_of(
-            f.args,
-            [] (typecheck::func_arg_t const& arg)
-            {
-                auto const term = std::get_if<typecheck::func_arg_t::term_arg_t>(&arg.value);
-                return term and is_first_order_type(term->type);
-            });
-}
-
-bool is_first_order_application(typecheck::expr_t::app_t const& f)
-{
-    return std::ranges::all_of(f.args, [] (auto const& x) { return ast::is_type(x.properties.sort); });
+    return is_first_order_type(f.ret_type.get()) and
```

Comment:
Should probably check arg types first to make sure that bound variables refer to first order types